pkg/systemd/init.js

```diff
@@ -0,0 +1,745 @@
+     *
+     * - The "Names" property of a unit only includes those aliases
+     *   that are currently loaded, not all.  To get all possible
+     *   aliases, one needs to call ListUnitFiles and match units via
+     *   their object path.
+     *
+     * - The unit file state of a alias as returned by ListUnitFiles
+     *   is always the same as the unit file state of the primary unit
+     *   file.
+     *
+     * - However, the unit file state as returned by ListUnitFiles is
+     *   not necessarily the same as the UnitFileState property of a
+     *   loaded unit.  ListUnitFiles reflects the state of the files
+     *   on disk, while a loaded unit is only updated to that state
+     *   via an explicit Reload.
+     *
+     * - Thus, we are careful to only use the UnitFileState as
+     *   returned by ListUnitFiles or GetUnitFileState.  The
+     *   alternative would be to only use the UnitFileState property,
+     *   but we need one method call per unit to get them all for the
+     *   overview, which seems excessive.
+     *
+     * - Methods like EnableUnitFiles only change the state of files
+     *   on disk.  A Reload is necessary to update the state
+     *   of loaded units.
+     *
+     * - A Reload will emit UnitRemoved/UnitNew signals for all units,
+     *   and no PropertiesChanges signal for the properties that have
+     *   changed because of the reload, such as UnitFileState.
+     *
+     */
+
+    function startsWith(string, prefix) {
+        return string.indexOf(prefix) === 0;
+    }
+
+    /* See systemd-escape(1), used for instantiating templates.
+     */
+
+    function systemd_escape(str) {
+
+        function name_esc(str) {
+            var validchars = /[0-9a-zA-Z:-_.\\]/;
+            var res = "";
+            var i;
+
+            for (i = 0; i < str.length; i++) {
+                var c = str[i];
+                if (c == "/")
+                    res += "-";
+                else if (c == "-" || c == "\\" || !validchars.test(c)) {
+                    res += "\\x";
+                    var h = c.charCodeAt(0).toString(16);
+                    while (h.length < 2)
+                        h = "0" + h;
+                    res += h;
+                } else
+                    res += c;
+            }
+            return res;
+        }
+
+        function kill_slashes(str) {
+            str = str.replace(/\/+/g, "/");
+            if (str.length > 1)
+                str = str.replace(/\/$/, "").replace(/^\//, "");
+            return str;
+        }
+
+        function path_esc(str) {
+            str = kill_slashes(str);
+            if (str == "/")
+                return "-";
+            else
+                return name_esc(str);
+        }
+
+        if (str.length > 0 && str[0] == "/")
+            return path_esc(str);
+        else
+            return name_esc(str);
+    }
+
+    var systemd_client = cockpit.dbus("org.freedesktop.systemd1", { superuser: true });
```

Comment:
The 'superuser: true' is only necessary on older systemd. Perhaps have a TODO here linking to our superuser optional bug?